dds/DCPS/DataWriterImpl_T.h

```diff
@@ -469,7 +470,11 @@ DataWriterImpl_T
           TraitsType::type_name()));
         return 0;
       }
-
+      if (encapsulated && (encoding.kind() == Encoding::KIND_XCDR1
+        || encoding.kind() == Encoding::KIND_XCDR2)) {
+        unsigned char padding = (tmp_mb->wr_ptr() - wr) % 4;
+        wr[serializer.swap_bytes() ? 0 : 3] |= (padding & 0x03);
```

Comment:
```suggestion
        unsigned char padding = (tmp_mb->wr_ptr() - wr) % 4;
        wr[3] |= (padding & 0x03);
```
The encapsulation header is a series of bytes, it's not effected by endianness.

Also please move this to logic a new function inside `EncapsulationHeader`.

Reply:
Unless I am reading this wrong, encapsulation header serialization is byte order sensitive:

ACE_INLINE void
Serializer::write_array(const char* x, size_t size, ACE_CDR::ULong length)
{
  this->write_array(x, size, length, this->swap_bytes_);
}

ACE_INLINE void
Serializer::write_array(const char* x, size_t size,
                        ACE_CDR::ULong length, bool swap)
{
  if (!swap || size == 1) {
    //
    // No swap, copy direct.
    //
    this->buffer_write(x, size * length, false);

  } else {
    //
    // Swapping _must_ be done at 'size' boundaries, so we need to spin
    // through the array element by element.
    // NOTE: This assumes that there is _no_ padding between the array
    //       elements.  If this is not the case, do not use this
    //       method.
    //
    while (length-- > 0) {
      this->buffer_write(x, size, true);
      x += size;
    }
  }
}

Reply:
That code is for an array of `unsigned long`. The encapsulation header uses an `octet` array:

https://github.com/objectcomputing/OpenDDS/blob/045bb54025689ce8b2aeb1017968f1a73708fbf1/dds/DCPS/Serializer.cpp#L244-L249

Which doesn't rely on endianness:

https://github.com/objectcomputing/OpenDDS/blob/045bb54025689ce8b2aeb1017968f1a73708fbf1/dds/DCPS/Serializer.inl#L671-L676

Reply:
Sorry no, it actually looks like the code you posted is for generic arrays, and encap header is using it. Size of  `byte_cdr_size` is one, so it's skipping doing anything about swapping because it doesn't apply.

Reply:
OK, that makes sense

Reply:
Do you prefer this line be within the EncupsulationHeader member function or handled by the caller?
`unsigned char padding = (tmp_mb->wr_ptr() - wr) % 4;`

Reply:
Yes, at least the padding part of it. Also I just realized this expression is wrong. It is the number of bytes the value is past the alignment, not the number of bytes to the next or current alignment. Please use: `align(value, 4) - value`. That will get the padding.

Reply:
Do I get the `value` as 
`size_t value = serialized_size(encoding, instance_data);`
or this also would be past the alignment?

Reply:
That depends on how we want to do it (see the discussion in the ticket), but `serialized_size` + the header size theoretically should be the same as `mb->wr_ptr() - wr`. Serialized size and the stream position I talked about are more "proper" than pointer arithmetic, but this is simpler.

Reply:
So neither `mb->wr_ptr() - wr` nor `serialized_size + the header size` gives us what we need, as both numbers include padding  - correct?

Reply:
Yes, but that is the whole point of this, is to mark how my bytes would be needed to bring the sample into 4 byte alignment.